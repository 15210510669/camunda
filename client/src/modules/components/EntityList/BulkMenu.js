/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React, {useState, useEffect, useRef} from 'react';

import {Dropdown, Deleter} from 'components';
import {t} from 'translation';
import {withErrorHandling} from 'HOC';
import {showError} from 'notifications';

export function BulkMenu({bulkActions, selectedEntries, onChange, mightFail}) {
  const bulkRef = useRef();
  const selectionMode = selectedEntries.length > 0;
  const [action, setAction] = useState(null);
  const [conflict, setConflict] = useState(false);
  const deleteAction = bulkActions.find((action) => action.type === 'delete');

  // remove the primary state of the main button when bulk actions button exists
  useEffect(() => {
    const button = bulkRef.current?.parentNode.querySelector('.action .Button.main');
    const togglePrimary = () => button?.classList.toggle('primary', !selectionMode);
    if (button) {
      togglePrimary();
      const menuObserver = new MutationObserver(togglePrimary);
      menuObserver.observe(button, {attributes: true});
      return () => {
        menuObserver.disconnect();
      };
    }
  }, [selectionMode]);

  function reset() {
    setConflict(false);
    setAction(null);
  }

  return (
    <div ref={bulkRef} className="BulkMenu">
      {selectionMode && (
        <Dropdown
          main
          primary
          label={`${selectedEntries.length} ${t(
            'common.itemSelected.' + (selectedEntries.length > 1 ? 'label-plural' : 'label')
          )}`}
        >
          {deleteAction && (
            <Dropdown.Option onClick={() => setAction(deleteAction)}>
              {t('common.delete')}
            </Dropdown.Option>
          )}
        </Dropdown>
      )}

      <Deleter
        type="items"
        entity={action?.type === 'delete' && selectedEntries}
        deleteEntity={() =>
          mightFail(
            action.action(selectedEntries),
            () => {
              reset();
              onChange();
            },
            showError
          )
        }
        onClose={reset}
        deleteButtonText={t('common.delete')}
        descriptionText={
          <>
            {conflict && (
              <>
                {action.conflictMessage}
                <br /> <br />
              </>
            )}
            {t('common.deleter.areYouSureSelected')}
          </>
        }
        checkConflicts={action?.checkConflicts}
        onConflict={() => setConflict(true)}
      />
    </div>
  );
}

export default withErrorHandling(BulkMenu);
