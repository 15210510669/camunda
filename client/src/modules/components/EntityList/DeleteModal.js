import React from 'react';

import {Button, Modal} from 'components';

export default function DeleteModal(props) {
  const {isVisible, entityName, deleteEntity, closeModal} = props;
  return (
    <Modal open={isVisible} onClose={closeModal} onConfirm={deleteEntity}>
      <Modal.Header>Delete {entityName}</Modal.Header>
      <Modal.Content>
        <p>You are about to delete {entityName}. Are you sure you want to proceed?</p>
      </Modal.Content>
      <Modal.Actions>
        <Button className="deleteModalButton" onClick={closeModal}>
          Close
        </Button>
        <Button
          type="primary"
          color="red"
          className="deleteEntityModalButton"
          onClick={deleteEntity}
        >
          Delete
        </Button>
      </Modal.Actions>
    </Modal>
  );
}
