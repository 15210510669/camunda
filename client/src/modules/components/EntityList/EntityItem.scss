.item {
  display: flex;
  border-bottom: 1px solid #ededed;

  &:first-child {
    border-top: 1px solid #ededed;
  }

  &:nth-child(even) {
    background: rgba(0, 0, 0, 0.03);
  }

  &:not(.noEntities):hover {
    background: rgba(0, 0, 0, 0.05);
  }

  &.noEntities {
    font-style: italic;
    display: flex;
    flex-direction: column;
    justify-content: center;
    text-align: center;
    padding: 12px 0;
    height: 148px;

    .createLink {
      margin-top: 14px;
    }
  }

  .info {
    padding: 12px 0;
    display: flex;
    flex-grow: 1;
    text-decoration: none;
    align-items: center;

    .visualizationIcon {
      padding: 10px;
      border-right: 1px solid #ededed;
      margin-right: 10px;
    }

    .textInfo {
      display: flex;
      width: 100%;
      margin-right: 20px;
      justify-content: space-between;
      flex-direction: column;

      .dataTitle {
        font-weight: bold;
        font-size: 18px;
        text-decoration: none;
        color: #333333;
        margin-bottom: 7px;
      }

      .extraInfo {
        display: flex;
        min-height: 30px;

        .dataMeta {
          font-size: 14px;
          color: #333333;
        }

        .dataIcons {
          max-width: 50px;
        }

        .custom {
          overflow: visible;
          color: #333333;
          font-size: 14px;
          flex-grow: 0;
          flex-shrink: 0;
          flex-basis: 60%;
        }
      }
    }
  }

  .operations {
    padding: 12px 0;
    display: flex;
    flex-shrink: 0;
    align-self: center;

    .dataTool {
      > * {
        width: 32px;
        height: 32px;
        border-radius: 100%;
        margin-right: 9px;
        display: flex;
        flex-direction: column;
        justify-content: center;
        align-items: center;
        border: 1px solid #dadee3;
        cursor: pointer;
        color: #6a727e;
        &:hover {
          background: #d6d6d6;
        }
      }

      a {
        color: #6a727e;
      }

      .editLink {
        font-size: 0.875em;
      }
    }
  }

  .data {
    padding: 0 4px;
  }
}
