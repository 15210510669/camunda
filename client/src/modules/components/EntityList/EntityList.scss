@import 'shared-styles/shared-styles.scss';

.EntityList {
  width: 100%;
  height: 100%;
  padding: 0 65px;
  overflow: auto;

  .empty {
    margin-top: 50px;
    margin-bottom: 20px;
    text-align: center;
    font-size: 14px;
    color: $grey-lighten-56;
  }

  &.scrolled .header {
    border-bottom: 1px solid $silver-darken-80;
    box-shadow: 0 2px 5px 0 rgba(0, 0, 0, 0.1);
  }

  &.embedded {
    padding: 0;

    .header {
      margin: 0;
      padding: 18px 0;
      width: 100%;

      .titleBar h1 {
        font-weight: bold;
        font-size: 16px;
        color: initial;
        line-height: normal;
        align-self: flex-end;
      }
    }
  }

  .header {
    margin: 0 -65px;
    padding: 0 65px;
    width: calc(100% + 130px);
    position: sticky;
    padding-top: 32px;
    top: 0;
    background-color: white;
    box-shadow: none;
    border-bottom: 1px solid transparent;
    transition: box-shadow 0.2s, border 0.2s;
    z-index: var(--z-above);

    .titleBar {
      line-height: 32px;
      display: flex;
      padding-bottom: 18px;

      h1 {
        font-size: 24px;
        line-height: 32px;
        font-weight: normal;
        color: $grey-darken-23;
        margin: 0;
        flex-grow: 1;
      }

      > .action:not(:empty) {
        margin-left: 20px;
      }

      .sortMenu {
        width: 32px;
        height: 32px;
        margin-left: 20px;
        margin-top: 2px;

        .activateButton .Icon {
          font-size: 24px;
        }
        .downIcon {
          display: none;
        }
        .menu {
          margin-top: -6px;
        }
      }
    }
  }

  .columnHeaders {
    display: flex;
    padding-bottom: 11px;
    padding-left: 97px;
    padding-right: 61px;
    color: $grey-lighten-56;
    font-size: 14px;
    line-height: 18px;

    > div {
      text-overflow: ellipsis;
      overflow: hidden;
      white-space: nowrap;
      padding-right: 10px;

      &.sortable > span {
        cursor: pointer;
      }

      &.sorted {
        color: $grey-darken-23;
      }

      .Icon {
        font-size: 24px;
        position: absolute;
        margin-left: 4px;
        margin-top: -3px;
        transition: 0.2s transform;

        &.desc {
          transform: rotate(180deg);
        }
      }
    }

    .name {
      flex: 1;
    }

    .meta {
      flex: 0 0 200px;
    }
  }

  ul {
    padding: 0;
    margin: 0 0 20px 0;
    transition: 0.1s opacity;

    &.isLoading {
      opacity: 0.34;
      pointer-events: none;
    }
  }

  .LoadingIndicator {
    margin-top: 50px;
    margin-left: calc(50% - 70px);
    position: absolute;
  }
}
