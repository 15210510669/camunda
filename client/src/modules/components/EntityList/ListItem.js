/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import React from 'react';
import {Link} from 'react-router-dom';
import classnames from 'classnames';

import {Input, Icon, Tooltip} from 'components';

import ListItemAction from './ListItemAction';

import './ListItem.scss';

export default function ListItem({
  isSelected,
  onSelectionChange,
  data: {type, name, link, icon, meta = [], actions, warning},
  hasWarning,
  singleAction,
}) {
  const content = (
    <>
      <Icon type={icon} />
      <div className="name">
        <span className="type">{type}</span>
        <Tooltip content={name} overflowOnly>
          <span className="entity">{name}</span>
        </Tooltip>
      </div>
      {meta.map((content, idx) => (
        <Tooltip key={idx} content={content} overflowOnly>
          <div className="meta">{content}</div>
        </Tooltip>
      ))}
      {hasWarning && (
        <div className="warning">
          {warning && (
            <Tooltip content={warning} theme="dark" delay={0}>
              <Icon type="error" size="18" />
            </Tooltip>
          )}
        </div>
      )}
      <ListItemAction actions={actions} singleAction={singleAction} />
    </>
  );

  return (
    <li
      className={classnames('ListItem', {
        active: isSelected,
        selectable: actions?.length > 0,
      })}
    >
      <Input type="checkbox" checked={isSelected} onChange={onSelectionChange} />
      {link ? <Link to={link}>{content}</Link> : <div>{content}</div>}
    </li>
  );
}
