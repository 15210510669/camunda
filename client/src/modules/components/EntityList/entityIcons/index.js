/* eslint import/no-webpack-loader-syntax: 0 */

import alert from '-!svg-react-loader!./alert.svg';
import alerts from '-!svg-react-loader!./alerts.svg';
import dashboard from '-!svg-react-loader!./dashboard.svg';
import dashboards from '-!svg-react-loader!./dashboards.svg';
import CombinedReportBar from '-!svg-react-loader!./report-combined-bar-chart.svg';
import reportBar from '-!svg-react-loader!./report-bar-chart.svg';
import reportHeat from '-!svg-react-loader!./report-heatmap.svg';
import CombinedReportLine from '-!svg-react-loader!./report-combined-line-chart.svg';
import reportLine from '-!svg-react-loader!./report-line-chart.svg';
import reportNumber from '-!svg-react-loader!./report-number.svg';
import reportPie from '-!svg-react-loader!./report-pie-chart.svg';
import CombinedReportTable from '-!svg-react-loader!./report-combined-table.svg';
import reportTable from '-!svg-react-loader!./report-table.svg';
import reports from '-!svg-react-loader!./reports.svg';
import reportEmpty from '-!svg-react-loader!./report-empty.svg';

const icons = {
  alert: {
    header: {Component: alerts},
    generic: {Component: alert}
  },
  report: {
    header: {Component: reports},
    generic: {Component: reportEmpty},
    bar: {label: 'Bar Chart Report', Component: reportBar, CombinedComponent: CombinedReportBar},
    heat: {label: 'Heatmap Report', Component: reportHeat},
    line: {
      label: 'Area Chart Report',
      Component: reportLine,
      CombinedComponent: CombinedReportLine
    },
    number: {label: 'Number Report', Component: reportNumber},
    pie: {label: 'Pie Chart Report', Component: reportPie},
    table: {label: 'Table Report', Component: reportTable, CombinedComponent: CombinedReportTable}
  },
  dashboard: {
    header: {Component: dashboards},
    generic: {Component: dashboard}
  }
};

export default icons;
