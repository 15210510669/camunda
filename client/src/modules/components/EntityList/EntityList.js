/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React, {useState} from 'react';
import {Link} from 'react-router-dom';

import {t} from 'translation';
import {LoadingIndicator, Icon, Input, Dropdown} from 'components';

import ListItem from './ListItem';

import './EntityList.scss';

export default function EntityList({name, children, action, isLoading, data, empty}) {
  const [searchQuery, setSearchQuery] = useState('');

  const searchFilteredData = (data || []).filter(({name}) =>
    name.toLowerCase().includes(searchQuery.toLowerCase())
  );

  const isEmpty = !isLoading && (data || []).length === 0;
  const hasResults = searchFilteredData.length > 0;

  return (
    <div className="EntityList">
      <div className="header">
        <h1>{name}</h1>
        <div className="searchContainer">
          <Icon className="searchIcon" type="search" />
          <Input
            required
            type="text"
            className="searchInput"
            placeholder={t('home.search.name')}
            value={searchQuery}
            onChange={({target: {value}}) => setSearchQuery(value)}
            onClear={() => setSearchQuery('')}
          />
        </div>
        <div className="action">{action}</div>
      </div>
      <div className="content">
        {isLoading && <LoadingIndicator />}
        {isEmpty && <div className="empty">{empty}</div>}
        {!isEmpty && !hasResults && <div className="empty">{t('common.notFound')}</div>}
        {hasResults && (
          <ul>
            {searchFilteredData.map(
              ({className, link, icon, type, name, meta1, meta2, meta3, action, actions}, idx) => {
                const content = (
                  <>
                    <ListItem.Section className="icon">{icon}</ListItem.Section>
                    <ListItem.Section className="name">
                      <div className="type">{type}</div>
                      <div className="entityName" title={name}>
                        {name}
                      </div>
                    </ListItem.Section>
                    <ListItem.Section className="meta1">{meta1}</ListItem.Section>
                    <ListItem.Section className="meta2">{meta2}</ListItem.Section>
                    <ListItem.Section className="meta3">{meta3}</ListItem.Section>
                  </>
                );

                return (
                  <ListItem key={idx} className={className} onClick={action}>
                    {link ? <Link to={link}>{content}</Link> : content}
                    {actions && actions.length && (
                      <div className="contextMenu" onClick={evt => evt.stopPropagation()}>
                        <Dropdown label={<Icon type="overflow-menu-vertical" size="24px" />}>
                          {actions.map(({action, icon, text}, idx) => (
                            <Dropdown.Option onClick={action} key={idx}>
                              <Icon type={icon} /> {text}
                            </Dropdown.Option>
                          ))}
                        </Dropdown>
                      </div>
                    )}
                  </ListItem>
                );
              }
            )}
          </ul>
        )}
        {children}
      </div>
    </div>
  );
}
