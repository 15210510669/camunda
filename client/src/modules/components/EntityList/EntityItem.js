import React from 'react';
import moment from 'moment';
import classnames from 'classnames';

import {Icon, Button} from 'components';
import {Link} from 'react-router-dom';
import entityIcons from './entityIcons';
import {formatters} from 'services';

import './EntityItem.css';

export default class EntityItem extends React.Component {
  getEntityIcon = entity => {
    let Icon;
    let label;

    const isValidCombined = this.isValidCombinedReport(entity);
    const iconKey = this.getIconKey(entity, isValidCombined);
    const iconData = entityIcons[this.props.api][iconKey];

    if (isValidCombined) {
      Icon = iconData.CombinedComponent;
      label = `Combined ${iconData.label}`;
    } else {
      Icon = iconData.Component;
      label = iconData.label;
    }

    const EntityIconComponent = label ? (
      <span title={label}>
        <Icon />
      </span>
    ) : (
      <Icon />
    );

    return EntityIconComponent;
  };

  getIconKey = (entity, isValidCombined) => {
    // if combined get the visualization type of the containing reports from the EntityList
    if (isValidCombined) return this.props.getReportVis(entity.data.reportIds[0]);
    if (entity.data && entity.data.visualization) return entity.data.visualization;
    return 'generic';
  };

  isValidCombinedReport = entity => {
    return (
      entity.reportType &&
      entity.reportType === 'combined' &&
      entity.data.reportIds &&
      entity.data.reportIds.length
    );
  };

  formatData = entity => {
    const {name, id, lastModified, lastModifier, shared} = entity;
    const entry = {
      name,
      link: `/${this.props.api}/${id}`,
      icon: this.getEntityIcon(entity),
      infos: [
        {
          parentClassName: 'custom',
          content: this.props.renderCustom && this.props.renderCustom(entity)
        },
        {
          parentClassName: 'dataMeta',
          content: (
            <React.Fragment>
              {`Last modified ${moment(lastModified).format('lll')} by `}
              <strong>{lastModifier}</strong>
            </React.Fragment>
          )
        },
        {
          parentClassName: 'dataIcons',
          content: shared && <Icon type="share" title={`This ${this.props.label} is shared`} />
        }
      ],
      operations: [],
      editData: entity
    };

    if (this.props.operations.includes('edit')) {
      entry.operations.push({
        content: <Icon type="edit" title={`Edit ${this.props.api}`} className="editLink" />,
        link: `/${this.props.api}/${id}/edit`,
        editData: entity,
        parentClassName: 'dataTool'
      });
    }

    if (this.props.operations.includes('duplicate')) {
      entry.operations.push({
        content: (
          <Button onClick={this.triggerDuplicate} noStyle>
            <Icon
              type="copy-document"
              title={`Duplicate ${this.props.api}`}
              className="duplicateIcon"
            />
          </Button>
        ),
        parentClassName: 'dataTool'
      });
    }

    if (this.props.operations.includes('delete')) {
      entry.operations.push({
        content: (
          <Button onClick={this.triggerDeleteModal} noStyle>
            <Icon type="delete" title={`Delete ${this.props.api}`} className="deleteIcon" />
          </Button>
        ),
        parentClassName: 'dataTool'
      });
    }

    return entry;
  };

  triggerDuplicate = evt => {
    evt.target.blur();
    this.props.duplicateEntity(this.props.data.id);
  };

  triggerDeleteModal = evt => {
    const {id, name} = this.props.data;
    return this.props.showDeleteModal({id, name});
  };

  renderCells = data => {
    return data.map((cell, idx) => (
      <span key={idx} className={classnames('data', cell.parentClassName)}>
        {this.renderCell(cell)}
      </span>
    ));
  };

  renderCell = cell => {
    if (cell.link) {
      return this.renderLink(cell);
    }
    return cell.content;
  };

  // if a ContentPanel is provided add onClick event otherwise use a router Link
  renderLink = data => {
    const {ContentPanel} = this.props;
    const allEntityData = data.editData;
    const EntityLink = ContentPanel ? 'a' : Link;
    const linkProps = {
      to: ContentPanel ? undefined : data.link,
      onClick: ContentPanel ? () => this.props.updateEntity(allEntityData) : undefined,
      className: data.className
    };
    return (
      <EntityLink {...linkProps}>
        {data.title ? (
          <React.Fragment>
            <span className="data visualizationIcon">{data.icon}</span>
            <div className="textInfo">
              <div className="data dataTitle">
                <h3>{formatters.getHighlightedText(data.title, this.props.query)}</h3>
                {allEntityData.reportType && allEntityData.reportType === 'combined' ? (
                  <span>Combined</span>
                ) : (
                  ''
                )}
              </div>
              <div className="extraInfo">{data.content}</div>
            </div>
          </React.Fragment>
        ) : (
          data.content
        )}
      </EntityLink>
    );
  };

  render() {
    const row = this.formatData(this.props.data);

    return (
      <li className="item">
        {this.renderLink({
          title: row.name,
          icon: row.icon,
          content: this.renderCells(row.infos),
          link: row.link,
          className: 'info',
          editData: row.editData
        })}
        <div className="operations">{this.renderCells(row.operations)}</div>
      </li>
    );
  }
}

EntityItem.defaultProps = {
  operations: ['create', 'edit', 'delete']
};
