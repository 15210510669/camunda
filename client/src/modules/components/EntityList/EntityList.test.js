/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a commercial license.
 * You may not use this file except in compliance with the commercial license.
 */

import React, {runLastEffect} from 'react';
import {shallow} from 'enzyme';
import update from 'immutability-helper';

import {Dropdown} from 'components';

import EntityList from './EntityList';

const props = {
  name: 'EntityList Name',
  empty: 'Empty Message',
  children: <div>Some additional Content</div>,
  action: <button>Click Me</button>,
  data: [
    {
      id: 'aCollectionId',
      name: 'aCollectionName',
      meta: ['Some info', 'Some additional info', 'Some other info'],
      icon: 'iconType',
      type: 'Collection',
      actions: [{icon: 'edit', text: 'Edit', action: jest.fn()}],
    },
    {
      id: 'aDashboardId',
      name: 'aDashboard',
      meta: ['Some info', 'Some additional info', 'Some other info'],
      icon: 'iconType',
      type: 'Dashboard',
    },
    {
      id: 'aReportId',
      name: 'aReport',
      meta: ['Some info', 'Some additional info', 'Some other info', 'special info'],
      icon: 'iconType',
      type: 'Report',
      link: 'link/to/somewhere',
    },
  ],
};

it('should match snapshot', () => {
  const node = shallow(<EntityList {...props} />);

  expect(node).toMatchSnapshot();
});

it('should show a loading indicator', () => {
  const node = shallow(<EntityList isLoading />);

  expect(node.find('LoadingIndicator')).toExist();
});

it('should show an empty message if no entities exist', () => {
  const node = shallow(<EntityList {...props} data={[]} />);

  expect(node.find('.empty')).toExist();
});

it('should filter results based on search input', () => {
  const node = shallow(<EntityList {...props} />);
  node.find('SearchField').simulate('change', 'adashboard');
  expect(node.find('ListItem').length).toBe(1);

  node.find('SearchField').simulate('change', 'special info');
  expect(node.find('ListItem').length).toBe(1);
});

it('should show no result found text when no matching entities were found', () => {
  const node = shallow(<EntityList {...props} />);

  node.find('SearchField').simulate('change', 'not found entity');

  expect(node.find('.empty')).toIncludeText('No results found');
});

it('should pass a hasWarning prop to all ListEntries if one of them has a warning', () => {
  const warningData = update(props.data, {0: {$merge: {warning: 'some warning'}}});

  const node = shallow(<EntityList {...props} data={warningData} />);

  node.find('ListItem').forEach((node) => expect(node.prop('hasWarning')).toBe(true));
});

it('should show a column header if specified', () => {
  const node = shallow(<EntityList {...props} columns={['Name', 'Meta 1', 'Meta 2', 'Meta 3']} />);

  expect(node.find('.columnHeaders')).toMatchSnapshot();
});

it('should hide search for embedded entitylist', () => {
  const node = shallow(<EntityList {...props} embedded />);

  expect(node.find('SearchField')).not.toExist();
});

it('should have a sorting menu if at least one column is sortable', () => {
  const node = shallow(
    <EntityList {...props} columns={[{name: 'Name', key: 'name', defaultOrder: 'asc'}, 'Meta 1']} />
  );

  expect(node.find('.sortMenu')).toExist();
});

it('should always call onsorting change with default order from the sorting menu', () => {
  const spy = jest.fn();
  const node = shallow(
    <EntityList
      {...props}
      columns={[{name: 'Name', key: 'name', defaultOrder: 'asc'}, 'Meta 1']}
      sorting={{key: 'name', order: 'asc'}}
      onChange={spy}
    />
  );

  node.find('.sortMenu').find(Dropdown.Option).simulate('click');

  expect(spy).toHaveBeenCalledWith('name', 'asc');
});

it('should not show a header entry for column entries that are hidden', () => {
  const node = shallow(
    <EntityList
      {...props}
      columns={['Name', {name: 'hidden column', hidden: true}, 'Another Column']}
    />
  );

  expect(node.find('.columnHeaders')).toMatchSnapshot();
});

it('should indicate which columns are sortable', () => {
  const node = shallow(
    <EntityList
      {...props}
      columns={['Name', {name: 'sortable', key: 'sortKey', defaultOrder: 'asc'}, 'Another Column']}
    />
  );

  expect(node.find('.columnHeaders div').at(1)).toHaveClassName('sortable');
  expect(node.find('.columnHeaders div').at(0)).not.toHaveClassName('sortable');
});

it('should indicate which column is sorted', () => {
  const node = shallow(
    <EntityList
      {...props}
      columns={[
        'Name',
        {name: 'sortable', key: 'sortKey', defaultOrder: 'asc'},
        {name: 'Another Column', key: 'sortKey2', defaultOrder: 'desc'},
      ]}
      sorting={{key: 'sortKey2', order: 'asc'}}
    />
  );

  expect(node.find('.columnHeaders div').at(1)).not.toHaveClassName('sorted');
  expect(node.find('.columnHeaders div').at(2)).toHaveClassName('sorted');
});

it('should call reload when clicking on the header column', () => {
  const spy = jest.fn();
  const node = shallow(
    <EntityList
      {...props}
      columns={[{name: 'sortable', key: 'sortKey', defaultOrder: 'asc'}]}
      onChange={spy}
    />
  );

  node.find('.columnHeaders span').simulate('click');

  expect(spy).toHaveBeenCalledWith('sortKey', 'asc');
});

it('should reverse the order when clicking on a header column that is already sorted', () => {
  const spy = jest.fn();
  const node = shallow(
    <EntityList
      {...props}
      columns={[{name: 'sortable', key: 'sortKey', defaultOrder: 'asc'}]}
      sorting={{key: 'sortKey', order: 'asc'}}
      onChange={spy}
    />
  );

  node.find('.columnHeaders span').simulate('click');

  expect(spy).toHaveBeenCalledWith('sortKey', 'desc');
});

it('should select and deselect a list item', () => {
  const node = shallow(<EntityList {...props} />);

  node
    .find('ListItem')
    .at(0)
    .simulate('selectionChange', {target: {checked: true}});

  expect(node.find({isSelected: true})).toExist();

  node
    .find('ListItem')
    .at(0)
    .simulate('selectionChange', {target: {checked: false}});

  expect(node.find({isSelected: true})).not.toExist();
});

it('should select/deselect all selectable events in view', () => {
  const customProps = {
    ...props,
    data: [
      props.data[0],
      {...props.data[1], actions: [{icon: 'delete', text: 'Delete'}]},
      props.data[2],
    ],
  };

  const node = shallow(<EntityList {...customProps} columns={['Name']} />);

  node
    .find('.columnHeaders')
    .find({type: 'checkbox'})
    .simulate('change', {target: {checked: true}});

  expect(node.find({isSelected: true}).length).toBe(2);

  node
    .find('.columnHeaders')
    .find({type: 'checkbox'})
    .simulate('change', {target: {checked: false}});

  expect(node.find({isSelected: true})).not.toExist();
});

it('should should reset the selection on data change', () => {
  const node = shallow(
    <EntityList
      {...props}
      columns={[
        'Name',
        {name: 'sortable', key: 'sortKey', defaultOrder: 'asc'},
        {name: 'Another Column', key: 'sortKey2', defaultOrder: 'desc'},
      ]}
    />
  );

  node
    .find('.columnHeaders')
    .find({type: 'checkbox'})
    .simulate('change', {target: {checked: true}});

  runLastEffect();

  expect(node.find({isSelected: true})).not.toExist();
});
