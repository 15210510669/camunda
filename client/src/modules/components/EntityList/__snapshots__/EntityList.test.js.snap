// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div
  className="EntityList"
  onScroll={[Function]}
>
  <div
    className="header"
  >
    <div
      className="titleBar"
    >
      <h1>
        EntityList Name
      </h1>
      <SearchField
        onChange={[Function]}
        value=""
      />
      <div
        className="action"
      >
        <button>
          Click Me
        </button>
      </div>
    </div>
  </div>
  <div
    className="content"
  >
    <ul
      className="itemsList"
    >
      <ListItem
        data={
          Object {
            "actions": Array [
              Object {
                "action": [MockFunction],
                "icon": "edit",
                "text": "Edit",
              },
            ],
            "icon": "iconType",
            "id": "aCollectionId",
            "meta": Array [
              "Some info",
              "Some additional info",
              "Some other info",
            ],
            "name": "aCollectionName",
            "type": "Collection",
          }
        }
        hasWarning={false}
        isSelected={false}
        key="0"
        onSelectionChange={[Function]}
        singleAction={true}
      />
      <ListItem
        data={
          Object {
            "icon": "iconType",
            "id": "aDashboardId",
            "meta": Array [
              "Some info",
              "Some additional info",
              "Some other info",
            ],
            "name": "aDashboard",
            "type": "Dashboard",
          }
        }
        hasWarning={false}
        isSelected={false}
        key="1"
        onSelectionChange={[Function]}
        singleAction={true}
      />
      <ListItem
        data={
          Object {
            "icon": "iconType",
            "id": "aReportId",
            "link": "link/to/somewhere",
            "meta": Array [
              "Some info",
              "Some additional info",
              "Some other info",
              "special info",
            ],
            "name": "aReport",
            "type": "Report",
          }
        }
        hasWarning={false}
        isSelected={false}
        key="2"
        onSelectionChange={[Function]}
        singleAction={true}
      />
    </ul>
    <div>
      Some additional Content
    </div>
  </div>
</div>
`;

exports[`should not show a header entry for column entries that are hidden 1`] = `
<div
  className="columnHeaders"
>
  <ForwardRef(Input)
    checked={false}
    className=""
    onChange={[Function]}
    type="checkbox"
  />
  <Tooltip
    content="Name"
    key="0"
    overflowOnly={true}
  >
    <div
      className="name"
    >
      <span
        onClick={[Function]}
      >
        Name
      </span>
    </div>
  </Tooltip>
  <Tooltip
    content="Another Column"
    key="1"
    overflowOnly={true}
  >
    <div
      className="meta"
    >
      <span
        onClick={[Function]}
      >
        Another Column
      </span>
    </div>
  </Tooltip>
</div>
`;

exports[`should show a column header if specified 1`] = `
<div
  className="columnHeaders"
>
  <ForwardRef(Input)
    checked={false}
    className=""
    onChange={[Function]}
    type="checkbox"
  />
  <Tooltip
    content="Name"
    key="0"
    overflowOnly={true}
  >
    <div
      className="name"
    >
      <span
        onClick={[Function]}
      >
        Name
      </span>
    </div>
  </Tooltip>
  <Tooltip
    content="Meta 1"
    key="1"
    overflowOnly={true}
  >
    <div
      className="meta"
    >
      <span
        onClick={[Function]}
      >
        Meta 1
      </span>
    </div>
  </Tooltip>
  <Tooltip
    content="Meta 2"
    key="2"
    overflowOnly={true}
  >
    <div
      className="meta"
    >
      <span
        onClick={[Function]}
      >
        Meta 2
      </span>
    </div>
  </Tooltip>
  <Tooltip
    content="Meta 3"
    key="3"
    overflowOnly={true}
  >
    <div
      className="meta"
    >
      <span
        onClick={[Function]}
      >
        Meta 3
      </span>
    </div>
  </Tooltip>
</div>
`;
